csharp/Ice/Middleware/Server/AuthorizationMiddleware.cs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) ZeroC, Inc.
+
+namespace Server;
+
+/// <summary>The AuthorizationMiddleware intercepts incoming requests and checks the token entry in their contexts.
+/// </summary>
+internal class AuthorizationMiddleware : Ice.Object
+{
+    private readonly Ice.Object _next;
+    private readonly string _validToken;
+
+    /// <inheritdoc />
+    public ValueTask<Ice.OutgoingResponse> dispatchAsync(Ice.IncomingRequest request)
+    {
+        // Check if the request has a valid token in its context.
+        if (!request.current.ctx.TryGetValue("token", out string? token) || token != _validToken)
+        {
+            Console.WriteLine($"Rejecting request with token '{token}'");
+
+            // ObjectNotExistException is one of the 6 Ice local exceptions that Ice transmits "over the wire".
+            // It usually means "no servant was found for this identity". Here, we reuse it to indicate an authorization
+            // failure without leaking information to the client.
+            throw new Ice.ObjectNotExistException();
```

Comment:
The most debatable aspect to this demo. There is no perfect solution with the Ice protocol since "system exceptions" are not extensible.

Defining a new user exception, like in the existing interceptor demo, is not good as it pollutes all Slice operation signatures.

UnknownException works but triggers by default a log from the logger middleware (Ice.Warn.Dispatch=1 by default). ONE or perhaps OpNE sounds like the best option.